### ISSUE TITLE
难以理解的一点

### ISSUE THREAD

**Issue by gaoren002** [cqylunlun/GLASS#49]
关于文中梯度引导的操作，我认为是个不错的idea，但是无论从论文描述还是代码描述我都无法理解，这里引导的方式，代码如下
``
true_loss = torch.nn.BCELoss()(true_scores, torch.zeros_like(true_scores))
                gaus_loss = torch.nn.BCELoss()(gaus_scores, torch.ones_like(gaus_scores))
                bce_loss = true_loss + gaus_loss
                grad = torch.autograd.grad(gaus_loss, [gaus_feats])[0]
                grad_norm = torch.norm(grad, dim=1)
                grad_norm = grad_norm.view(-1, 1)
                grad_normalized = grad / (grad_norm + 1e-10)
                with torch.no_grad():
                    gaus_feats.add_(0.001 * grad_normalized)

我能明白这里使用梯度上升是为了远离正常区域，但是梯度源自gaus_loss = torch.nn.BCELoss()(gaus_scores, torch.ones_like(gaus_scores))而不是gaus_loss = torch.nn.BCELoss()(gaus_scores, torch.zeros_like(gaus_scores))，我比较难以理解

**Comment by gaoren002** [comment]
就代码的实际效果而言，我认为实际上是逼近了正常区域，而非远离了正常区域

**Comment by cqylunlun** [comment]
感谢您的认可。基于梯度上升的合成异常的确逼近了正常区域，从而实现了模型对于微弱缺陷的判别学习。该步骤的目的是为了合成更加贴近于正常区域的异常特征，因此标签为1。更加详细的理论推导过程，请参考[DROCC: Deep Robust One-Class Classification](https://proceedings.mlr.press/v119/goyal20c.html)。

**Comment by gaoren002** [comment]
但是论文所述似乎是说更加远离尤其是论文插图中所提到tsne图？我主要对此感到迷惑。

**Comment by gaoren002** [comment]
而且实际上我认为梯度引导的方向不是最重要的，重要的是不能让合成异常的本质（高斯噪声）被模型学习到，实现这一点有一种思路是利用扩散模型在隐空间用高斯噪声进行扰动。我认为梯度引导同样实现了这一点

**Comment by cqylunlun** [comment]
> 但是论文所述似乎是说更加远离尤其是论文插图中所提到tsne图？我主要对此感到迷惑。

文中的解释确实容易引起歧义，tsne图是判别学习后的特征分布，GAS合成异常已经可以与正常区域区分开来。

**Comment by cqylunlun** [comment]
> 而且实际上我认为梯度引导的方向不是最重要的，重要的是不能让合成异常的本质（高斯噪声）被模型学习到，实现这一点有一种思路是利用扩散模型在隐空间用高斯噪声进行扰动。我认为梯度引导同样实现了这一点

这是个很有启发性的思路，高斯噪声还可以继续深挖。

**Comment by gaoren002** [comment]
> > 但是论文所述似乎是说更加远离尤其是论文插图中所提到tsne图？我主要对此感到迷惑。
> 
> 文中的解释确实容易引起歧义，tsne图是判别学习后的特征分布，GAS合成异常已经可以与正常区域区分开来。
好吧，但是文章还有很多文字图像表述，比如Fig3，都在暗示是远离正常样本的方向在引导

**Comment by cqylunlun** [comment]
可能存在表述上的混淆。以图3为例，它是一种抽象的示意，强调高斯噪声扰动后，利用截断投影将特征约束在合适范围内，避免离正常区域过近或过远。
